pkg/controller/common/helpers.go

```diff
@@ -306,30 +306,36 @@ func ValidateMachineConfig(cfg mcfgv1.MachineConfigSpec) error {
 // IgnParseWrapper parses rawIgn for both V2 and V3 ignition configs and returns
 // a V2 or V3 Config or an error. This wrapper is necessary since V2 and V3 use different parsers.
 func IgnParseWrapper(rawIgn []byte) (interface{}, error) {
-	ignCfg, rpt, err := ign2.Parse(rawIgn)
-	if err == nil && !rpt.IsFatal() {
-		// this is an ignv2cfg that was successfully parsed
-		return ignCfg, nil
-	}
-	if err.Error() == ign2error.ErrUnknownVersion.Error() {
-		ignCfgV3, rptV3, errV3 := ign3.Parse(rawIgn)
-		if errV3 == nil && !rptV3.IsFatal() {
-			return ignCfgV3, nil
+	ignCfgV3, rptV3, errV3 := ign3.Parse(rawIgn)
+	if errV3 == nil && !rptV3.IsFatal() {
+		return ignCfgV3, nil
+	}
+	// unlike spec v2 parsers, v3 parsers aren't chained by default so we need to try parsing as spec v3.0 as well
+	if errV3.Error() == ign3error.ErrUnknownVersion.Error() {
+		ignCfgV3_0, rptV3_0, errV3_0 := ign3_0.Parse(rawIgn)
+		if errV3_0 == nil && !rptV3_0.IsFatal() {
+			return translate3.Translate(ignCfgV3_0), nil
 		}
-		// unlike spec v2 parsers, v3 parsers aren't chained by default so we need to try parsing as spec v3.0 as well
-		if errV3.Error() == ign3error.ErrUnknownVersion.Error() {
-			ignCfgV3_0, rptV3_0, errV3_0 := ign3_0.Parse(rawIgn)
-			if errV3_0 == nil && !rptV3_0.IsFatal() {
-				return translate3.Translate(ignCfgV3_0), nil
+
+		if errV3_0.Error() == ign3error.ErrUnknownVersion.Error() {
+			ignCfgV2, rptV2, errV2 := ign2.Parse(rawIgn)
+			if errV2 == nil && !rptV2.IsFatal() {
+				return ignCfgV2, nil
 			}
 
-			return ign2types.Config{}, errors.Errorf("parsing Ignition config spec v3.0 failed with error: %v\nReport: %v", errV3_0, rptV3_0)
-		}
+			// If the error is still UnknownVersion it's not a 3.1/3.0 or 2.x config, thus unsupported
+			if errV2.Error() == ign2error.ErrUnknownVersion.Error() {
+				return ign3types.Config{}, errors.Errorf("parsing Ignition config failed: unknown version:"+
+					"v3 error %v\nv3 Report: %v\nv2 error %v\nv2 Report: %v\n", errV3, rptV3, errV2, rptV2)
+			}
 
-		return ign2types.Config{}, errors.Errorf("parsing Ignition config spec v3.1 failed with error: %v\nReport: %v", errV3, rptV3)
+			// Parsing all 3 types failed, let's return both 3.1 and 2.x errors
+			return ign3types.Config{}, errors.Errorf("parsing Ignition config failed: could not parse config in v2 or v3,"+
+				"v3 error %v\nv3 Report: %v\nv2 error %v\nv2 Report: %v\n", errV3, rptV3, errV2, rptV2)
+		}
+		return ign3types.Config{}, errors.Errorf("parsing Ignition config spec v3.0 failed with error: %v\nReport: %v", errV3, rptV3)
```

Comment:
are we treating 3.0/3.1 as completely different? if we support 3.2 will we need another return statement? maybe just v3?

Reply:
I will double check this

Reply:
Yeah so basically if we have 3.2 we will have to chain parse with another if clause, this is a ignition design decision